tools/OMIcheck/OMIcheck.ps1

```diff
@@ -1,127 +1,460 @@
+    }
+}
+
+# Find if the VMs in running state and has Linux OS.
+function IsVMRunningLinux($VM)
+{
+    $provisioningState = $VM.Statuses | where { $_.Code.StartsWith("ProvisioningState") }
+    if ($provisioningState.Code -ne "ProvisioningState/succeeded")
+    {
+        Write-Host -ForegroundColor Gray `t`t $VM.Name ": VM is not fully provisioned"
+        return $false
+    }
+    $powerState = $VM.Statuses | where { $_.Code.StartsWith("PowerState") }
+    if ($powerState.Code -ne "PowerState/running")
+    {
+        Write-Host -ForegroundColor Gray `t`t $VM.Name ": VM is not running"
+        return $false
+    }
+
+    if ($VMitem -ne $null -and $VMitem.StorageProfile.OsDisk.OsType.ToString() -ne "Linux")
+    {
+        Write-Host -ForegroundColor Gray `t`t $VM.Name ": VM is not running Linux OS"
+        return $false
+    }
+
+    return $true
+}
+
+# Update the VM extensions and OMI package in a subscription.
+# TODO: This can be converted to ForEach-Object -Parallel. However, it requires PSh >= 7.1
+function UpdateVMs($sub)
+{
+    # Set Azure Subscription context
+    Set-AzContext -Subscription $sub.Id
+
+    Write-Output "Listing Virutal Machines in subscription '$($sub.Name)'"
+    $VMs = Get-AzVM
+    $VMsSorted = $VMs | Sort-Object -Property ResourceGroupName
+    $PreviousRG = ""
+    foreach($VMitem in $VMs)
+    {
+        $VM = Get-AzVM -Name $VMitem.Name -ResourceGroupName $VMitem.ResourceGroupName -Status
+
+        # DEBUG: limit to 1 vm only
+        #if ($VM.name -ne "my-testing-vm")
+        #{
+        #    continue
+        #}
+
+        # Group VMs by resource groups.
+        if ($PreviousRG -ne $VM.ResourceGroupName)
+        {
+            Write-Host "`tResourceGroup: " $VM.ResourceGroupName
+            $PreviousRG = $VM.ResourceGroupName
+        }
+
+        if (!(IsVMRunningLinux($VM)))
+        {
+            continue
+        }
+
+        $check = IsVMVulnerableOMI($VM)
+        if ($check -eq $null)
+        {
+            Write-Host -ForegroundColor Gray `t`t $VM.Name ": VM has no OMI package. Skipping. "
+            continue
+        }
+
+        if ($check.OmiServerVer -eq $OmiServerGoodVersion)
+        {
+            Write-Host -ForegroundColor Green `t`t $VM.Name ": VM has patched OMI version " $check.OmiPkgVer
+            continue
+        }
 
-            # Update only running VMs.
-            if ($VM.PowerState -ne "VM running")
+        Write-Host -ForegroundColor Red `t`t $VM.Name ": VM has vulnerable OMI version " $check.OmiPkgVer
+        if ($upgradeOMI)
+        {
+            if ($check.LadPkgVer -ne $LadPkgGoodVersion)
             {
-                Write-Host -ForegroundColor Gray `t`t $VM.Name " in resource group : VM is not running"
-                continue
+                UpdateLAD($VM)
             }
 
-            # Update Linux only VMs.
-            if ($VM.StorageProfile.OsDisk.OsType.ToString() -ne "Linux")
+            if ($check.OmsPkgVer -ne $OmsPkgGoodVersion)
```

Comment:
i thought we didn't want to upgrade OMS if we upgraded LAD already, in case of compat issues

Reply:
Given that now both LAD and OMS is available in all regions; it makes sense to update both. Also, the delta in the hotfix is minor; so previous versions compat should apply. what do you think?